File: app/src/main/java/com/ytempest/wanandroid/activity/main/home/HomeFrag.java
package com.ytempest.wanandroid.activity.main.home;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;

import com.ytempest.banner.Banner;
import com.ytempest.banner.transformers.Transformers;
import com.ytempest.layoutinjector.annotation.InjectLayout;
import com.ytempest.wanandroid.R;
import com.ytempest.wanandroid.activity.main.home.article.HomeArticleAdapter;
import com.ytempest.wanandroid.base.fragment.LoaderFrag;
import com.ytempest.wanandroid.base.load.ViewType;
import com.ytempest.wanandroid.helper.ArticleDetailHelper;
import com.ytempest.wanandroid.http.bean.ArticleDetailBean;
import com.ytempest.wanandroid.http.bean.BannerBean;
import com.ytempest.wanandroid.http.bean.HomeArticleBean;

import java.util.List;

import butterknife.BindView;

/**
 * @author heqidu
 * @since 2020/6/23
 */
@InjectLayout(R.layout.frag_home)
public class HomeFrag extends LoaderFrag<HomePresenter> implements IHomeContract.View {

    private static final String TAG = HomeFrag.class.getSimpleName();

    @BindView(R.id.view_main_pull_refresh)
    SwipeRefreshLayout mRefreshLayout;
    @BindView(R.id.view_main_nested_scroll)
    NestedScrollView mNestedScrollView;
    @BindView(R.id.view_main_banner)
    Banner mBannerView;
    @BindView(R.id.rv_main_article_list)
    RecyclerView mRecyclerView;
    private HomeArticleAdapter mAdapter;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView();
        initData();
    }

    private void initView() {
        mRefreshLayout.setColorSchemeResources(R.color.main_color);
        mRefreshLayout.setOnRefreshListener(() -> mPresenter.refreshHomeArticle());

        mBannerView.setBannerBinder(new HomeBannerBinder());
        mBannerView.setPlayDuration(3000);
        mBannerView.setScrollDuration(1500);
        mBannerView.setScrollAnimation(Transformers.DAMPING);

        // 移除RecyclerView更新数据时的动画，防止View闪烁的问题
        mRecyclerView.setItemAnimator(null);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mAdapter = new HomeArticleAdapter(mPresenter);
        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.setNestedScrollingEnabled(true);
        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if (isArriveBottom()) {
                    mPresenter.loadMoreHomeArticle();
                }
            }
        });
    }

    private boolean isArriveBottom() {
        // RecyclerView.canScrollVertically(1)的值表示是否能向上滚动，false表示已经滚动到底部
        // RecyclerView.canScrollVertically(-1)的值表示是否能向下滚动，false表示已经滚动到顶部
        // 由于NestedScrollView嵌套了RecyclerView，这里需要通过NestedScrollView判断
        boolean isArriveBottom = !mNestedScrollView.canScrollVertically(1);
        boolean isIdleState = mRecyclerView.getScrollState() == RecyclerView.SCROLL_STATE_IDLE;
        return isArriveBottom && isIdleState;
    }

    private void initData() {
        getLoader().showView(ViewType.LOAD);
        mPresenter.loadHomeData();
        ArticleDetailHelper.getInstance().getArticleUpdateDetail().observe(getViewLifecycleOwner(), bean -> {
            if (bean == null || bean.getSource() != ArticleDetailBean.Source.HOME) return;
            for (HomeArticleBean.DatasBean data : mAdapter.getSrcDataList()) {
                if (data.getId() == bean.getArticleId()) {
                    data.setCollect(bean.isCollected());
                    mAdapter.refresh(data);
                    return;
                }
            }
        });
    }

    @Override
    protected void onReloadClick() {
        super.onReloadClick();
        mPresenter.loadHomeData();
    }

    @Override
    public void onHomeDataSuccess(List<BannerBean> bannerList, HomeArticleBean bean) {
        getLoader().hideAll();
        mBannerView.display(bannerList);
        mAdapter.display(bean.getDatas());
    }

    @Override
    public void onHomeDataFail(int code) {
        getLoader().showView(ViewType.ERR);
    }

    @Override
    public void displayHomeArticle(boolean fromRefresh, HomeArticleBean homeArticleBean) {
        if (fromRefresh) {
            mRefreshLayout.setRefreshing(false);
            mAdapter.display(homeArticleBean.getDatas());
        } else {
            mAdapter.addData(homeArticleBean.getDatas());
        }
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (hidden) {
            mBannerView.stopAutoPlay();
        } else {
            mBannerView.startAutoPlay();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mBannerView.startAutoPlay();
    }

    @Override
    public void onPause() {
        super.onPause();
        mBannerView.stopAutoPlay();
    }

    @Override
    public void onArticleCollectSuccess(HomeArticleBean.DatasBean data) {
        mAdapter.refresh(data);
    }

    @Override
    public void onArticleCollectFail(HomeArticleBean.DatasBean data, int code) {
        showToast(data.isCollect() ? R.string.collect_fail : R.string.cancel_fail);
    }
}

File: app/src/main/java/com/ytempest/wanandroid/activity/main/home/HomeFrag.kt
package com.ytempest.wanandroid.activity.main.home

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ytempest.wanandroid.R
import com.ytempest.wanandroid.activity.main.home.article.HomeArticleAdapter
import com.ytempest.wanandroid.base.createViewModel
import com.ytempest.wanandroid.base.fragment.MVVMFragment
import com.ytempest.wanandroid.base.load.Loader
import com.ytempest.wanandroid.base.load.ViewType
import com.ytempest.wanandroid.base.vm.EntityObserver
import com.ytempest.wanandroid.databinding.FragHomeBinding
import com.ytempest.wanandroid.helper.ArticleDetailHelper
import com.ytempest.wanandroid.http.bean.ArticleDetailBean
import com.ytempest.wanandroid.http.bean.BannerBean
import com.ytempest.wanandroid.http.bean.HomeArticleBean

/**
 * @author heqidu
 * @since 21-2-9
 */
class HomeFrag : MVVMFragment<FragHomeBinding>(), IHomeView {

    override val viewModel by lazy { createViewModel<HomeViewModel>() }
    private val mAdapter by lazy { HomeArticleAdapter(viewModel) }

    private val loader: Loader by lazy {
        Loader(binding.root as ViewGroup).also {
            it.setReloadCall {
                it.showView(ViewType.LOAD)
                viewModel.loadHomeData()
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
        initData()
    }

    private fun initView() {
        binding.refreshView.setColorSchemeResources(R.color.main_color)
        binding.refreshView.setOnRefreshListener { viewModel.refreshHomeArticle() }

        binding.bannerView.apply {
            setBannerBinder(HomeBannerBinder())
            setPlayDuration(3000)
            setScrollDuration(1500)
        }

        binding.articlesView.run {
            itemAnimator = null
            layoutManager = LinearLayoutManager(context)
            adapter = mAdapter
            isNestedScrollingEnabled = true
            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                    if (isArriveBottom()) {
                        viewModel.loadMoreHomeArticle()
                    }
                }
            })
        }
    }

    private fun isArriveBottom(): Boolean {
        // RecyclerView.canScrollVertically(1)的值表示是否能向上滚动，false表示已经滚动到底部
        // RecyclerView.canScrollVertically(-1)的值表示是否能向下滚动，false表示已经滚动到顶部
        // 由于NestedScrollView嵌套了RecyclerView，这里需要通过NestedScrollView判断
        val isArriveBottom = !binding.nestedScrollView.canScrollVertically(1)
        val isIdleState = binding.articlesView.scrollState == RecyclerView.SCROLL_STATE_IDLE
        return isArriveBottom && isIdleState
    }

    private fun initData() {
        loader.showView(ViewType.LOAD)
        ArticleDetailHelper.instance.getArticleUpdateDetail().observe(viewLifecycleOwner, Observer {
            if (it == null || it.source != ArticleDetailBean.Source.HOME) {
                return@Observer;
            }
            for (data in mAdapter.srcDataList) {
                if (data.id == it.articleId) {
                    data.collect = it.isCollected
                    mAdapter.refresh(data)
                }
            }
        })

        viewModel.homeDataResult.observe(this, EntityObserver(
                onSuccess = { entity ->
                    val (bannerList, bean) = entity.data
                    onHomeDataSuccess(bannerList, bean)
                },
                onFail = { entity ->
                    onHomeDataFail(entity.code)
                }
        ))

        viewModel.homeArticlesResult.observe(this, EntityObserver(
                onSuccess = { entity ->
                    val (fromRefresh, bean) = entity.data
                    displayHomeArticle(fromRefresh, bean)
                }
        ))

        viewModel.homeArticleCollectResult.observe(this, EntityObserver(
                onSuccess = { entity ->
                    onArticleCollectSuccess(entity.data)
                },
                onFail = { entity ->
                    onArticleCollectFail(entity.extra as HomeArticleBean.Data, entity.code)
                }
        ))

        viewModel.loadHomeData()
    }

    override fun onHomeDataSuccess(bannerList: List<BannerBean>, bean: HomeArticleBean) {
        loader.hideAll()
        binding.bannerView.display(bannerList)
        mAdapter.display(bean.datas)
    }

    override fun onHomeDataFail(code: Int) {
        loader.showView(ViewType.ERR)
    }

    override fun displayHomeArticle(fromRefresh: Boolean, homeArticleBean: HomeArticleBean) {
        if (fromRefresh) {
            binding.refreshView.isRefreshing = false
            mAdapter.display(homeArticleBean.datas)

        } else {
            mAdapter.addData(homeArticleBean.datas)
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (hidden) binding.bannerView.stopAutoPlay()
        else binding.bannerView.startAutoPlay()
    }

    override fun onResume() {
        super.onResume()
        binding.bannerView.startAutoPlay()
    }

    override fun onPause() {
        super.onPause()
        binding.bannerView.stopAutoPlay()
    }

    override fun onArticleCollectSuccess(data: HomeArticleBean.Data) {
        mAdapter.refresh(data)
    }


    override fun onArticleCollectFail(data: HomeArticleBean.Data, code: Int) {
        showToast(if (!data.collect) R.string.collect_fail else R.string.cancel_fail)
    }
}